src/xrc/src/lib.rs

```diff
@@ -402,22 +440,40 @@ impl AllocatedBytes for Vec<u64> {
     }
 }
 
-impl From<QueriedExchangeRate> for ExchangeRate {
-    fn from(rate: QueriedExchangeRate) -> Self {
-        ExchangeRate {
-            base_asset: rate.base_asset,
-            quote_asset: rate.quote_asset,
-            timestamp: rate.timestamp,
-            rate: median(&rate.rates),
-            metadata: ExchangeRateMetadata {
-                decimals: DECIMALS,
-                base_asset_num_queried_sources: rate.base_asset_num_queried_sources,
-                base_asset_num_received_rates: rate.base_asset_num_received_rates,
-                quote_asset_num_queried_sources: rate.quote_asset_num_queried_sources,
-                quote_asset_num_received_rates: rate.quote_asset_num_received_rates,
-                standard_deviation: standard_deviation(&rate.rates),
-                forex_timestamp: rate.forex_timestamp,
-            },
+impl TryFrom<QueriedExchangeRate> for ExchangeRate {
+    type Error = ExchangeRateError;
+
+    fn try_from(rate: QueriedExchangeRate) -> Result<Self, Self::Error> {
+        match rate.median_rate {
+            Some(median_rate) => {
+                // If no rates were invalid,the median over all rates can be used.
```

Comment:
The logic is the following: We only return a rate if the `median_rate` field is set. In other words, if it is not set, something went wrong, resulting in an error.
The `median_rate`, however, might actually be a function of multiple medians, for example the product of two medians.
This is not the same thing as the median over the product of rates. That's where the conditional check comes in: If all rates were valid, it is safe to compute the median over the product of rates for a more accurate result.

Reply:
But what guarantees that `median_rate` that is picked in line 455 is valid? (in the case the condition does not hold)

Reply:
There is code in the constructor to ensure that `median_rate` is valid (i.e., more than half of the received values are valid). If rates are dropped, we use the product of two medians (which are both valid individually) instead of the median of the cross product of rates.

Reply:
Sorry, I still don't get something here (and other than convincing me, it might be good to make the comment clear enough to convince other future readers)
If `median_rate` is ensured to be valid, why don't we just use it in all cases?

Reply:
I can explain the rationale face-to-face but the design is likely to change quite a bit because of the compatibility problem that @dfinity-ryancroote raised.

If we remove `median_rate` again, there is no need to add some elaborate comments now.   :-)

Reply:
The median_rate is now gone. The new mechanism should be easier to understand. @yotamhc, please take another look!